### PR TITLE
Added sealing of functions in traits

### PR DESCRIPTION
**Pull request by TheLazyDutchman**
This PR adds the ability to seal specific functions.

This is done through adding the `#[seal]` attribute on the function.
This makes the entire function private.

If you want the function to be callable but not implementable, you need to add `callable` to the `seal` attribute.

Since sealing the trait is semantically somewhat equivalent to adding `#[seal(callable)]` to all its functions.

To allow users more freedom, I added a `partial` argument to the trait attribute.
This attribute removes the seal from the trait itself, so that the user has almost full freedom in what functions are implementable and what are callable.

## Possible future additions
- Allow users to specify the name of the inner version of a wrapped function. 
- Allow users to specify the name of the token type that is used to make functions sealed.
- Add use statement for the token, to make it nicer to call the sealed function.

### CODE REVIEW

**Comment by TheLazyDutchman** [comment]
fixes #18

File: src/lib.rs [hunk 1]
```diff
@@ -124,171 +124,401 @@ use proc_macro::TokenStream;
...
+
+fn parse_function_arguments(
+	attrs: &[syn::Attribute],
+) -> Result<Option<(&syn::Attribute, FunctionArguments)>, syn::Error> {
+	for attr in attrs {
+		if attr.path().is_ident("seal") {
+			let value = if let syn::Meta::List(list) = &attr.meta {
+				syn::parse(list.tokens.clone().into())?
+			} else {
+				FunctionArguments::default()
```
Comment (jmg-duarte) [inline src/lib.rs:342]: Replace this with an error.

Let's force the users to write `#[seal(callable)]` or `#[seal(uncallable)]` so the attribute becomes more self-documenting

File: src/lib.rs [hunk 2]
```diff
@@ -124,171 +124,401 @@ use proc_macro::TokenStream;
...
+
+		match ident.to_string().as_str() {
+			"callable" => {
+				syn::Ident::parse(input)?;
+				out.callable = true
+			}
+			unknown => {
+				return Err(syn::Error::new(
+					ident.span(),
+					format!("unknown `{}` attribute argument", unknown),
```
Comment (jmg-duarte) [inline src/lib.rs:517]: Given we're going to force users to pick between `callable` and `uncallable`, explain which options are available. Something among the lines of:
```suggestion
					format!("unknown `{}` attribute argument. #[seal] only supports `callable` and `uncallable`", unknown),
```

File: src/lib.rs [hunk 3]
```diff
@@ -124,171 +124,401 @@ use proc_macro::TokenStream;
...
+	item_trait
+		.supertraits
+		.push(parse_quote!( #seal::Sealed #ty_generics ));
+
+	let mut wrappers = Vec::<syn::TraitItem>::new();
+
+	let mut all_implementable = true;
+	let mut none_implementable = true;
+
+	for item in &mut item_trait.items {
```
Comment (jmg-duarte) [inline src/lib.rs:164]: The new partial behavior represents a different approach to sealing, thus all the new behavior should be in its separate function, guarded by an `if args.partial`, you can extract the existing code to a separate function too in order to simplify code generation in the end.

To be more explicit, the macro should generate everything as it was until this change, and if `partial` is on, generate something else.

File: src/lib.rs [hunk 4]
```diff
@@ -124,171 +124,401 @@ use proc_macro::TokenStream;
...
+		.split_for_impl();
 
-    Ok(quote! {
-        #[automatically_derived]
-        impl #trait_generics #sealed_path #arguments for #self_type #where_clauses {}
-        #item_impl
-    })
+	for item in &mut item_impl.items {
+		if let syn::ImplItem::Fn(item) = item {
+			let input = parse_function_arguments(&item.attrs)?;
```
Comment (jmg-duarte) [inline src/lib.rs:297]: Can we: 
- make `FunctionArguments` consider the whole attribute (i.e. `#[seal(...)]`
- get the respective token stream and use `parse_macro_input!` 

That way we get, less code and more concise code

File: src/lib.rs [hunk 5]
```diff
@@ -124,171 +124,401 @@ use proc_macro::TokenStream;
...
+			return Ok(Some((attr, value)));
+		}
+	}
+
+	Ok(None)
+}
+
+/// Seal a specific function (to avoid people overwriting behaviour of
+/// specific functions, or even calling them)
+fn parse_function_definition(
```
Comment (jmg-duarte) [inline src/lib.rs:354]: Parsing _should not modify_ the function. Inline this code at the call site instead.

**Review by jmg-duarte — CHANGES_REQUESTED** [review]
I've done a first pass, it's looking good but there's still work to do.

Asides from the comments I left through the code, I ask you:

- Rebase/merge the new master into your changes, I've updated some things that will hopefully make developer experience more consistent.
- Replace new strings using line breaks with `concat!` — it formats much better and I find it more legible (especially in narrower windows).
- Add tests using both `partial` and `erased`.

**Comment by TheLazyDutchman** [comment]
One thing I am not happy about is that the way to make a sealed function callable is by adding a wrapper function that calls the sealed function.

But since this wrapper function is not sealed, and (in my current implementation) has the normal name and arguments. This means it is possible (in fact it is quite easy to do accidentally) to overwrite the wrapper that is used to call the sealed implementation.

**Comment by jmg-duarte** [comment]
@TheLazyDutchman I was reviewing your code and going through the examples when I noticed the following the following code does not work:

```rust
#[sealed(partial)]
pub trait A {
    #[seal(uncallable)]
    fn sealed() {}

    fn lol(&self) {
        Self::sealed();
    }
}
```

Now, it makes sense that it doesn't! The issue is, it needs to work. As a library author, you need to be able to write the previous code (with, of course, the `Self::sealed(Token)` instead). 

This works when done by hand because you're in full control of the names, but not with macros (as Rust doesn't have the [same abilities as F#](https://dl.acm.org/doi/pdf/10.1145/3178372.3179495)).

I see 3 possible ways out of this:
- Rewriting every call to the function to use the generated value — very convoluted
- Figuring out a syntax to make this approach work — possible but may be opaque to the end user, don't let this discourage you though!
- Documenting the new pattern in the macro — easiest possible solution

**Comment by TheLazyDutchman** [comment]
I think we can add doc attrs to document that token needs to be used, and then have a use statement for the token with the same publicity of the sealed module. This way you can just use `Token` instead of `_sealed_my_trait::Token` everywhere where you are allowed to use the sealed trait.

Since this directly puts `Token` in scope it means we should allow the user to specify their own name for token, to avoid name collisions. (or instead of `Token` we make it `MyTraitToken`)

**Comment by jmg-duarte** [comment]
I'm not being able to visualize your proposal. Can you show an example of how the final code would look like?

**Comment by TheLazyDutchman** [comment]
```
pub(super) _sealed_my_trait {
    pub trait Seal {}
    pub struct MyTraitToken;
}
pub(super) use _sealed_my_trait::MyTraitToken;
```

Where `pub(super)` is just to showcase what would happen if users specify a `pub` in the `sealed` attribute.

This doesn't change that users need to add the token, but at least they don't need to use the path of the module, and we can change the doc comments of the function to signify that they need the token in order to call it.

**Comment by jmg-duarte** [comment]
@TheLazyDutchman I wasn't too clear when asking for an example. I meant, how would an end-user write the code?

**Comment by TheLazyDutchman** [comment]
```rust
#[sealed]
pub trait MyTrait {
    #[seal(uncallable)]
    fn sealed() {}

    fn unsealed() {
        Self::sealed(Token);
    }
}
```
They still need to add the token, but we have a use statement for it, so `Token` is all they need, not the full path.
And we add to the documentation of the function that they need to supply the token.

**Comment by TheLazyDutchman** [comment]
I am not sure if it is possible to get the full semantics of a sealed function without adding an argument (at least that is what was done in the blog). The problem is that when doing it manually, you know what is going on, and this library would be hiding some of it (which is a good thing), but it cannot hide everything, which means it is unclear to the user where this token is coming from.

I think this can be somewhat fixed by adding documentation (both for this library, and for the function that we are sealing).

**Comment by jmg-duarte** [comment]
As far as VSCode is concerned, it doesn't have a parameter, which becomes really misleading for the user. 

Screenshot 2023-12-19 at 15 44 00

As you said:
> but it cannot hide everything, which means it is unclear to the user where this token is coming from.

I have to thank you for your efforts but I cannot move forward with this PR. I am open to merge a PR documenting the pattern from the blog though as it is indeed a really interesting pattern and still relevant to the people interested in this crate.

**Comment by TheLazyDutchman** [comment]
I was also unsatisfied with the `Token` trick and the wrapped function, and after thinking some more about it I think I have found a way to make a method be non-implementable but callable without using tokens.

But since this library normally already does exactly that for all methods, it would create two completely distinct behaviours, so I think it might be better as a separate library.

The new trick would be:
- create a sealed wrapper type `Wrapper<T>`
- `impl<T> From<T> for Wrapper<T>`
- change one (or all) argument from e.g `String` to `impl Into<Wrapper<String>>` (this requires at least one argument.
- add a line at the start of the function to rebind the argument name to the wrapped value.

**Pull request closed without merging**